ouroboros-consensus/test-consensus/Test/Dynamic/Network.hs

```diff
@@ -189,6 +196,19 @@ createNetworkInterface btime chans nodeIds joinSlotOf us
     maxSlot :: SlotNo -> SlotNo ->  SlotNo
     maxSlot (SlotNo i) (SlotNo j) = SlotNo (max i j)
 
+    -- handle a ForkTooDeep exception by restarting at the start of the next
+    -- slot
+    restartOnForkTooDeep :: forall a. m a -> m a
```

Comment:
TODO: will also be resolved when the test setup properly handles exceptions in connection threads.

Reply:
> when the test setup properly handles exceptions in connection threads

What do you mean by "properly handles"? In particular, the _restart after a delay_ behavior is key here, and I haven't found that aspect present anywhere else in the code base. (Maybe via `Ouroboros.Network.Subscription`, but it's not clear to me from looking at the code that that addresses the same issue.)

Reply:
In particular, I haven't open an Issue or PR for "proper exception handling in connection threads". I didn't follow the brief Slack discussion well enough to write one -- I see a couple trees but not the forest. If you have a sense of it, would you open an Issue for us to reference here?

Reply:
Related observation: the current commit restarts the ChainSync client on `ForkTooDeep` and `InvalidRollForward`. The client throws these only when handling `IntersectFound` and `RollForward` respectively from the server. In both cases, the client-server pair would have ended up in the `Idle` state according to `network.pdf` if the exceptions weren't thrown. So I think it's indeed a safe for the client to fail and then restart when handling these messages: a fresh client starts in the `Idle` state, which is what the server expects after sending those messages.

The server has no indication that the previous messages it sent were not successfully handled, but I don't think it matters; the protocol is stateless in that respect IIUC. In fact, maybe "restarting" is indeed "successfully handling" the message.

Reply:
> > when the test setup properly handles exceptions in connection threads
> 
> What do you mean by "properly handles"? In particular, the _restart after a delay_ behavior is key here, and I haven't found that aspect present anywhere else in the code base. (Maybe via `Ouroboros.Network.Subscription`, but it's not clear to me from looking at the code that that addresses the same issue.)

I mean the following: in connection threads (BlockFetch/ChainSync/TxSubmission Client/Server), many different exceptions can be thrown:
* Something fatal to the node occurred, e.g., a file-system error, in which case the whole node should be shut down, as opposed to simply restarting the respective connection thread.
* The other end of the connection exhibits adversarial behaviour, e.g., it sent an invalid block. In this case, we should terminate all connections to that node (and blacklist it?).
* We're temporarily not interested in the node, e.g., the `ForkTooDeep` exception. In this case, we terminate the respective connection temporarily (a configurable timeout), but we leave the other direction (i.e., `ChainSyncServer`) and the other connections running.

@coot is working on implementing this, see #903, #939, etc.

As these policies affect consensus, we should of course include them in the consensus tests. How we'll do this exactly, we still need to figure out. Ideally, the error policy implementation is parameterised over the actual connections, so we can simply plug in our test `Channel`s. How we'll deal with the configuration of the timeouts, is still an open question to me (we could just decide on the actual timeout to use in practice, and use that for the tests).